### DISCUSSION TITLE
Memory Management

### DISCUSSION THREAD

**Discussion by directtosumit** [linuxmint/discussions#201] [Ideas]
I have been using Linux mint for almost 4 to 5 years, not just for casual usage, but for hardcore software development like Android, React Native, HTML, python,c# etc., But the main problem which I have faced mostly when the system runs out of memory, and the system just freezes, I can't even move my mouse, and it is dead until I restart my laptop, and everything is lost, it is very bad user experience.

I have a suggestion, please select some softwares which are essential for the system to keep going, then always allocate some part of ram memory for those software so that system does not ever freeze, for example if I have 8 GB of ram, I am very happy to provide 2 GB all the time for system to keep running, now if a process freezes and uses up rest of 6 GB then I can easily kill the process and get back my memory.

**Comment by Jeremy7701** [comment]
You need to identify the particular software that is a memory hog.
Is this a new problem?

How do you know its a memory problem since you have an apparent system freeze?
Does `sudo journalctl -b -1 -p3` (for example) show anything useful when run in the reboot following a freeze error?

**Reply by directtosumit** [reply]
I will use that command next time whenever my laptop freezes.

**Comment by directtosumit** [comment]
Simply because, I know exactly which programs are causing it,
While working with react native, and trying to install the Android app, the
Java program uses lots of memory, I frequently check how much memory is
left with "free -h" command, and exactly when my ram memory is just about
to be over, the system becomes dead, if somehow I am able kill any program
before it becomes completely unresponsive, like chrome, or sublime or
intellij Idea, or any other program, immediate system recovers some memory
and laptop gets back to it's normal state.

**Comment by Jeremy7701** [comment]
Since you know the exact cause of the problem and I assume that allocating a lot of swap space doesn't help, then the only solution is a memory upgrade.

Alternatively, terminate unnecessary large programs which aren't needed before trying to simultaneously "While working with react native, and trying to install the Android app, the Java program uses lots of memory" .
 `journalctl` is basically a log of warning/error messages issued by various components.

**Reply by Reno-Sifana** [reply]
Hi. after i install Linux (e.g Debian, Ubuntu, Linux Mint and others) i have install necessary programs (Including nohang an OOM Killer https://github.com/hakavlad/nohang). In My Opinion Nohang is good OOM Killer. Nohang is not aggressive killing program than systemd-oomd
For Your Information: Nohang is featured in Garuda Linux (An Arch Linux Based Distro).

**Comment by directtosumit** [comment]
So you are basically saying to buy a better laptop with better ram support,
that much sense I have, and I would buy a better laptop when I can, but I
wanted to change a fundamental thing on next Linux mint upgrade, no matter
how much memory I put, ultimately if too much programs are open my laptop
will freeze, and this is a fundamental problem, 3-4 years of suffering, you
should not just overturn with a suggestion - "buy a better laptop" that I
already know.

**Reply by ghost** [reply]
How about earlyoom? I just started using it due to Flightgear having an issue were if it crashes it goes full on memory leak and has to be killed because the usual "would you like to close the unresponsive possess" will think it is responding when it is not.

Anyway the earlyoom description is " Earlyoom is an userspace OOM-killer which can avoid the system to get into unresponsive state caused by swapping, which is most likely to occur when a large swap is present and memory is tight. It checks the amount of available memory and swap periodically, and when both are below a preconfigured value, it kills the largest process."

**Reply by ghost** [reply]
https://github.com/orgs/linuxmint/discussions/1052

**Comment by directtosumit** [comment]
Sometimes I just need a minute to kill the Java program and start again,
but since the system is completely unresponsive I can not do anything but
restarting my laptop.

**Comment by Jeremy7701** [comment]
I appreciate the problem: but running multiple very large programs, especially Java is going to use a lot of RAM.

Since you are considering killing a program, is it not possible to run a background script which can monitor free memory and perform the kill process automatically?
Obviously this isn't ideal!

**Comment by directtosumit** [comment]
You give me such a script, I could not find any, why not incorporated such
a script in the system itself, and temporarily freeze the programs and
inform the user that system will soon run out of memory, please take some
action immediately, rather than just being completely dead, does it make
sense to you ?

**Comment by Jeremy7701** [comment]
Good idea - a bit like the warning you get when your approaching 100% of disk usage.

My programming skills are fairly basic - so I'd run a BASH script in a terminal window.
In this setup, you need an active window to display the message in.
You can use the `killall` command to suspend / resume / terminate **known programs** - you don't need to know its process-id.
The option to freezing everything you are authorised to freeze (possible) doesn't sound great.

This sounds a bit like a problem I once had with a video transcoding program trying to run many threads at 100% causing the CPU to trip out due to the temperature. So I monitored the temperature and got killall to send alternate STOP/CONT signals. I copied bits of it from the internet.
You might find the zenity package useful - [https://www.howtogeek.com/107537/how-to-make-simple-graphical-shell-scripts-with-zenity-on-linux/](https://www.howtogeek.com/107537/how-to-make-simple-graphical-shell-scripts-with-zenity-on-linux/)

Here's my example - not a good one, I admit!!!
```
#! /bin/bash

if [[ $# -eq 1 ]]; then
   pgm="$1"
else
   pgm="ghb"
fi
ACTIVE="YES"

while [[ 0 -ne 1 ]]; do
sleep 10

HOT=$(cat /sys/class/hwmon/hwmon0/temp1_input | cut -b 1-2)

if [[ $ACTIVE == "YES" ]] && [[ $HOT -gt 60 ]]; then
    killall -s STOP "$PGM"
    ACTIVE="NO"
fi

if [[ $ACTIVE == "NO" ]] && [[ $HOT -lt 55 ]]; then
    killall -s CONT "$PGM"
    ACTIVE="YES"
fi
done
```
Obviously, you would need to extract the free memory perhaps using the `free` command & issue an appropriate message.

**Comment by anaximeno** [comment]
Try to install and configure ZRAM, it might help a bit with freezes

**Comment by AtomicRobotMan0101** [comment]
Agree with @anaximeno on the `ZRAM`.  It is wise.

Also, rather than killing the machine, hit `CTRL-ALT-F1` to `F6`

This will toss you into a TTY in which you can log in, run `top` or `htop` and find which program is misbehaving and `kill <pid>` or `renice` it.

Once it is back under control, `CTRL-ALT-F7` brings us back into the cinnamon-land :)

Using the `CTRL-ALT-F1` to `F6` trick is fundamental to the kernel and sit underneath everything. If that is borked, well, you are doomed.

Also, rather than turning off the machine, give it a chance to restart gracefully with `ALT-PRTSCR` then type `REISUB`.  Leave a moment, perhaps 10 or 15 seconds before pressing the `B`.

REISUB issues commands directly to the kernel. REISU essentially cuts everything down, B is Boot....

......

Editing - I thought to add this for the sake of completeness and for those following in future: https://wiki.archlinux.org/title/Keyboard_shortcuts#Kernel_(SysRq)

There is also a section just below it which deals with OOM and how the use of `ALT-PRTSCR-f` will have the effect OP  @directtosumit seeks: https://wiki.archlinux.org/title/Keyboard_shortcuts#Killing_a_memory-hogging_process

**Comment by Jeremy7701** [comment]
It's a chunk of software for one person.
I never see OOM, because I don't try to run a bunch of giant clunky process simultaneously.
Running a background process to check for something that I'm never going to see is pointless.

Have you considered a more lightweight distro which doesn't use too much RAM?
